htlcswitch/link.go

```diff
@@ -2272,16 +2279,27 @@ func (l *channelLink) CheckHtlcTransit(payHash [32]byte,
 	policy := l.cfg.FwrdingPolicy
 	l.RUnlock()
 
-	return l.canSendHtlc(
-		policy, payHash, amt, timeout, heightNow,
-	)
+	err := l.canSendHtlc(policy, payHash, amt, timeout, heightNow)
+	if err == nil {
+		return nil
+	}
+
+	return err.message
 }
 
-// htlcSatifiesPolicyOutgoing checks whether the given htlc parameters satisfy
-// the channel's amount and time lock constraints.
+// linkFailure represents failure to forward a HTLC along the link. It contains
+// an appropriate lnwire message and the specific reason for failure.
+type linkFailure struct {
+	message    lnwire.FailureMessage
```

Comment:
Using `ForwardingError` and using existing `lnwire` errors would be good, since there are a lot of duplicates. What about defining the extra errors in the `htlcnotifier` (rather than `lnwire`) package which implement `FailureMessage` and then mapping custom errors back to their broadcast-friendly error? Eg, insufficient balance -> temp channel failure.